caffe2/core/registry.h

```diff
@@ -1,6 +1,11 @@
 /**
  * Simple registry implementation in Caffe2 that uses static variables to
  * register object creators during program initialization time.
+ *
+ * WARNING: this registry is not entirely thread-safe, as reads to
+ * the registry are not protected by a mutex.  The safest mode of use
+ * is to dlopen() *all* dynamic libraries that may write to the library
+ * and synchronize prior to performing any reads on the registry.
```

Comment:
Also, doesn't it have to be `dlopen` with `RTLD_GLOBAL`? Is `RTLD_LAZY` enough to ensure all initializers are called?

Reply:
Per the manpage:

```
RTLD_LAZY
Perform lazy binding. Only resolve symbols as the code that references them is executed. If the symbol is never referenced, then it is never resolved. (Lazy binding is only performed for function references; references to variables are always immediately bound when the library is loaded.)
```

Registerers are static variables (not functions) and so they are immediately bound.